docs/adr/0007-seperate-binding-generators.md

```diff
@@ -0,0 +1,142 @@
+# Separate the binding generators into their own crates per target
+
+* Status: proposed
+* Deciders: TBD
+* Date: 2022-01-16
+
+Technical Story: [Issue 299](https://github.com/mozilla/uniffi-rs/issues/299)
+
+outdated early prototype: [PR 997](https://github.com/mozilla/uniffi-rs/pull/997)
+
+## Context and Problem Statement
+All the binding generators currently live in the [`uniffi_bindgen`](../../uniffi_bindgen/src/bindings) crate. This creates the following difficulties:
+- All the bindings live in the `uniffi` repository, so the `uniffi` team has to maintain them (or at the very least review changes to them).
+- Any changes to a specific binding generator require a new `uniffi_bindgen` release for it to be released. Even if it doesn't impact any of the other bindings.
+- Some bindings require complex build systems to test, and including those build systems in `uniffi` adds its complexity. For example, any type of `gecko-js` bindings would require the mozilla-central build system to build and test.
+- We currently run all the tests for the bindings in our CI and through `cargo test`. This means that if one binding target gets outdated and fails, or if a developer doesn't have the needed libraries installed for one of the targets, the tests would fail.
+
+This ADR re-evaluates the architecture of how we consume the bindings in hope of improving the developer experience of adding new bindings, testing new and the current bindings and simplifying the ownership of the bindings.
+## Decision Drivers
+
+* Maintainability, we would like to improve the maintainability of our binding generators - this means clear ownership of the bindings.
+* Testability, it should be easy for developers to test the bindings they care about. Without having to navigate and install unfamiliar libraries and build systems.
```

Comment:
I think there are two sides to this.  We want to make it easy to ignore languages that we don't care about (Ruby, maybe Python), but we want to make it easy to test all the languages we do (Kotlin and Swift).  I'm hoping that whatever solution we end up with, it's easy to run the test fixtures for all languages I care about with a single command.

Reply:
I would appreciate if we _do_ care about Python. Glean's gonna be a user of that.

Reply:
Thanks, @mhammond and @bendk!! This brings up something important. I generally agree with accepting changes that break bindings (even if they are the core ones) - I like the idea of using semver to make sure the versions are not coupled... but one thing we'll have to think carefully about while making this decision is whether that forces us to move the bindings we have in this repository away.

Once we establish that it's okay to break bindings, and bindings should depend on uniffi versions for which they are not broken, the bindings that live in this repository should then depend on a concrete version of uniffi - as opposed to a path dependency.

assuming they stay in this repository, I can think of a couple of implications:
- Releases would have to be done separately for the uniffi libraries and the uniffi_bindgen_lang crates, this makes it hard to reason about what github releases on this repository mean (i.e do we even cut github releases for bindings crates? maybe this is OK?)
- CI could start to be complicated, as it would test each binding crate, possibly using a completely different version of uniffi (which could be OK? can't concretely think of a big drawback to this except that CI might have to download different versions of uniffi)
 
I can open another issue to discuss if we should move the current binding generator crates in this repo to another repository (because that has its own implications and tradeoffs). 

But in the meantime, I'm curious what you think about the semver question given the above, are the implications worth it? or should we just say "no, **for now,** you shouldn't break any binding that lives in this repository because they are all tested and released together with `uniffi`"?

Reply:
> just say "no, for now, you shouldn't break any binding that lives in this repository because they are all tested and released together with uniffi"?

Yes, I think that is the most pragmatic thing to do. I was really just thinking out-loud before - I don't think we should try and optimize for the case of people breaking bindings :) It does raise some interesting questions about policies regarding which bindings are special enough to be in the same repo, but I don't think we necessarily need to answer them now.